client/extensions/woocommerce/woocommerce-services/views/carrier-accounts/dynamic-settings-form.js

```diff
@@ -0,0 +1,199 @@
+
+/**
+ * Internal dependencies
+ */
+import Button from 'components/button';
+import CompactCard from 'components/card/compact';
+import Checkbox from 'woocommerce/woocommerce-services/components/checkbox';
+import TextField from 'woocommerce/woocommerce-services/components/text-field';
+import { getSelectedSiteId } from 'state/ui/selectors';
+import * as api from 'woocommerce/woocommerce-services/api';
+import { errorNotice, successNotice } from 'state/notices/actions';
+
+export const CheckboxFormFieldSet = ( props ) => {
+    return (
+        <>
+            <Checkbox
+                id={ props.fieldKey }
+                onChange={ props.onChange }
+                checked={ props.checked || false }
+            />
+            <label htmlFor={ props.fieldKey }><span>{ props.labelName }</span></label>
+        </>
+    );
+};
+
+export const DynamicCarrierAccountSettingsForm = ( props ) => {
+	const {
+        translate,
+    } = props;
+
+    const listOfFormFieldSet = [];
+
+    const [formFields, setFormFields] = useState({});
+    const [isSaving, setIsSaving] = useState(false);
+
+    const updateValue = ( fieldKey ) => ( newValue ) => {
+        const formFieldsCopy = {...formFields};
+        formFieldsCopy[fieldKey] = newValue;
+        setFormFields(formFieldsCopy);
+    }
+
+    const toggleCheckbox = ( fieldKey ) => () => {
+        const formFieldsCopy = {...formFields};
+        formFieldsCopy[fieldKey] = !formFieldsCopy[fieldKey];
+        setFormFields(formFieldsCopy);
+    };
+
+    const getInputComponent = (fieldKey) => {
+        const visibility = props.registrationFields[fieldKey].visibility;
+        const labelName = props.registrationFields[fieldKey].label;
+
+        let formComponent;
+
+        switch (visibility) {
+            case 'select':
+            case 'checkbox':
+                formComponent = <CheckboxFormFieldSet
+                    key={ fieldKey }
+                    fieldKey={ fieldKey }
+                    labelName={ labelName }
+                    onChange={toggleCheckbox(fieldKey)}
+                    checked={formFields[fieldKey]}
+                />
+
+                break;
+            case 'invisible':
+            case 'masked':
+            case 'readonly':
+                // TODO: We will need to handle the above invisible/readonly fields separately. For now, use default.
+            case 'password':
+                // TODO: We will need to create a PasswordField component, for now, use default.
+            case 'visible':
+            default:
+                formComponent = <TextField
+                    id={ fieldKey }
+                    key={ fieldKey }
+                    title={ labelName }
+                    updateValue={ updateValue( fieldKey ) }
+                    value={formFields[fieldKey]||""}
+                />
+                break;
+        }
+        return formComponent;
+    };
+
+    const submitCarrierSettingsHandler = async () => {
```

Comment:
Having an `async` handler can have some performance implications. This blocks the browser and "freezes" the UI until the promise is resolved.

I recommend using a little trick to prevent this: declare the async function _inside_ the handler and call it _from_ the handler.

An example is added to the other comment with all the code suggestions.

Reply:
Good callout! This reminds me of a eslint warning thrown by https://github.com/facebook/react/blob/369c3db6299506d9d1ab6628accc2c40827d1f1f/packages/eslint-plugin-react-hooks/src/ExhaustiveDeps.js#L103-L113. I didn't know it would freezes the UI. 

Maybe we should add eslint-plugin-react-hooks to our eslint rules. There are some useful things for hooks.